### ISSUE TITLE
ENH: Endianness should not be lost for `common_type`/array coercion and maybe ufunc-calls

### ISSUE THREAD

**Issue by nschloe** [numpy/numpy#15088]
When `column_stack`ing two arrays, the dtype of the result is the dtype of the input arrays (if they are are the same). However, if the endianness of the input array is not the system endianness, it is not respected.

MWE:
```python
import numpy

a = numpy.array([1, 2, 3], dtype="i4")
b = numpy.array([1, 2, 3], dtype="i4")
c = numpy.column_stack([a, b])
print(a.dtype)  # int32   okay
print(b.dtype)  # int32   okay
print(c.dtype)  # int32   okay
print()

a = numpy.array([1, 2, 3], dtype="=i4")
b = numpy.array([1, 2, 3], dtype="=i4")
c = numpy.column_stack([a, b])
print(a.dtype)  # int32   okay
print(b.dtype)  # int32   okay
print(c.dtype)  # int32   okay
print()

a = numpy.array([1, 2, 3], dtype="<i4")
b = numpy.array([1, 2, 3], dtype="<i4")
c = numpy.column_stack([a, b])
print(a.dtype)  # int32   okay
print(b.dtype)  # int32   okay
print(c.dtype)  # int32   okay
print()

a = numpy.array([1, 2, 3], dtype=">i4")
b = numpy.array([1, 2, 3], dtype=">i4")
c = numpy.column_stack([a, b])
print(a.dtype)  # >i4     okay
print(b.dtype)  # >i4     okay
print(c.dtype)  # int32   WRONG
print()
```

_Edit_:
Even more obvious:
```python
import numpy

a = numpy.array([1, 2, 3], dtype=">i4")

print(a.dtype)  # >i4
print((+a).dtype)  # int32 !
```

**Comment by seberg** [comment]
Yeah, as soon as a type promotion occurs, numpy will lose the endianness of the dtype(s) even if they are identical. I am a bit torn on this. In this case (promotion for "concatenation" like operations) that seems a bit strange. For ufuncs I am not sure if it actually makes sense to not do the `>int + >int -> >int` dance (although those could behave differently).

**Comment by seberg** [comment]
The bigger issue is whether we can/should change it considering that code may rely on it to some degree. Although, for most practical purposes it would just make code slow since it works with the incorrect byteorder.

**Comment by rossbar** [comment]
Is this behavior (losing endianness when a type promotion occurs) documented anywhere? I did a quick search through the docs for `dtype`, `promote_types`, and `result_type` and saw no mention of it. Is this the type of thing that should potentially raise a run time warning or do you think that would be overkill?

Note that `np.result_type` illustrates the observed behavior:
```python
import numpy as np
>>> a = np.array([1, 2, 3], dtype=">i4")
>>> b = np.array([4, 5, 6], dtype=">i4")
>>> np.result_type(a, b)
dtype('int32')
>>> np.result_type(a, b) == a.dtype
False
```

**Comment by seberg** [comment]
Yeah, I suppose that `result_type` at least should likely document it. Or maybe it should just be mentioned in a/the general paragraph on type promotion (i.e. we have the promotion tables somewhere).

**Comment by rossbar** [comment]
I was looking at it a little more and came up with the following case:

```python
>>> d = np.dtype('>i4')
>>> np.result_type(d)
dtype('>i4')
>>> np.result_type(d, d)
dtype('int32')
```
This output of `np.result_type` might be a little confusing because PyArray_ResultType "short-cuts" and just returns the input dtype when there is only one argument, without ever evaluating the type promotion. However, the type promotion *does* occur even for ufuncs with only one input, e.g.

```python
>>> a = np.array([1, 2, 3], dtype=">i4")
>>> b = np.positive(a)
>>> b.dtype
dtype('int32')

I still think the solution is just to document the behavior better. @seberg I propose following up on your suggestion and adding a warning the Casting Rules section of the ufuncs docs, as well as modifying the docstring for `np.result_type`. If this seems sensible I will create a PR.

**Comment by seberg** [comment]
Sounds good to me, thanks!

**Comment by nschloe** [comment]
Thanks @rossbar for your example. This is really pretty crazy:
```python
import numpy

a = numpy.array([1, 2, 3], dtype=">i4")

print(a.dtype)  # >i4
print((+a).dtype)  # int32 !
```
This means you can't do anything in nonnative byteorder; it gets rewritten immediately.

You do what you do, but for the record I need to add that I don't really like the documentation approach. I've seen it happening a number of times before in numpy/scipy: There's a quirky behavior, and instead of tackling it (which can be hard), it gets documented somewhere in a website. What's that good for? In the future of course you can just point towards that website and say: See, we even documented it this way! That doesn't make it any less quirky though. If anything, it carves this kind of behavior in stone because, well, "it's documented so people rely on this and we cannot change it". (I've heard this one too often before.) I almost feel like it's better _not_ to document it to give it the chance to be fixed in the future.

**Comment by seberg** [comment]
Well, fair enough. I am not against keeping an issue open (or anyone tackling it). But it is one of the things that are very tricky to judge who/how it affects people :(. I want to do more such changes, but there are higher ones on the list of priorities (ragged objects, value based promotion, windows integer behaviour, ...).

I would like to separate out the two things:

1. `common_type` (used for concatenate, etc.); And the behaviour during `np.array()` calls:
    * I actually agree, preserving endianess would be better
2. `promotion` in UFUncs: They actually always drop the endianess. I suppose the reason for that is that it is always produces native byte order. As you said, there is no way to get around this unless you specify an `out` argument (even the `signature=(None, arr.dtype)` will drop it – in my language, it uses the `DType(Type)`, in the form of the type number, and not the specific instance)

In any case, I think you are right and it is at the very least one of those things we should consider when we decide for a major release: I.e. its a hard to change break, but it also probably affects very few people.

**Comment by rossbar** [comment]
Thanks for the feedback - I'm perfectly happy to close #15100 so as not to discourage further discussion. There is a draft of a [proposed Numpy Enhancement Proposal](https://github.com/numpy/numpy/blob/a111b551ae940d7d5f8523fef1cf3589c6ba00a0/doc/neps/nep-0033-extensible-dtypes.rst) that explicitly mentions the casting system; maybe some discussion can be added in conjunction with that.

Is it maybe worthwhile to re-draft the issue (or create a new one) with the more general theme to encourage further discussion?

I can think of at least one use-case where this affects users: I've used `np.frombuffer` with structured dtypes to parse data from network packets (big-endian) in the past. That's one case where it would be really useful to know that byte order is being implicitly changed by the numpy machinery as soon as you actually try to operate on the data.

**Comment by seberg** [comment]
I think Nico is definetly right about the `common_type` (point 1) there is just no real logic/consistency in converting, only the convenience to not run into the trap of bad performance later. I am not 100% sure about the promotion in ufuncs.

I suppose the main issue is that working with non-native byteorder will be in general slow, so aside from people use from/to buffer (and being surprised by this or the change), people reading non-native byte-order may need to convert manually to not slow down their code drastically.

**Comment by charris** [comment]
I think possible use cases need to be carefully examined. I don't think emulating a machine with non-native byte order isn't something we should attempt.

**Comment by tkzv** [comment]
At the very least, automatic changing of endianness should be properly documented. When a user sets endianness explicitly, there should be an error, if the end result cannot have it.

**Comment by eric-wieser** [comment]
@tkzv's point is that `np.func(..., dtype=non-native)` should not return native-endianess data, which I think I agree with. @rkern is right that this is not very often something you want to do anyway, but if you're sure enough to ask for it explicitly, numpy probably ought to comply.

**Comment by seberg** [comment]
Just stumbled on this and I think the past decisions/comments to do nothing are correct.  (The way the `dtype` argument and `cumsum` works is indeed a different issue.)  Note that we also don't even have byte-order on scalars (gh-1866).

It seems, I previously thought that preserving things for `concatenate` has some merit.  But, I now think this is just a good sane – and simple – default to drop.  And I am not convinced that trying to modify it would improve anything.

In the end, I doubt we would ever be able to make strong, consistent, guarantees about preserving byte-order.  Trying to do so would add an unnecessary layer of complexity _and_ may even lead to a worse/more complicated user-experience.

*We just shouldn't make many strong guarantees about "storage details".*

Some long ramblings, best ignored ☺:

<details>

Byte-order (similar to e.g. encoding of strings) is a "storage detail" that has no effect on the value or behaviour of the "type" represented by the dtype.  There is usually one "canonical" version of dtypes: that is native byte-order.  And NumPy prefers that, and will convert to it silently.
I think this is fine: NumPy should not have to consider the storage detail as contracts, but be able to pick them as it thinks is best.  

Note that this is not very different from the fact that we modify memory order when it makes things faster/easier (we have some logic to preserve it often, but it is not particularly reliable).
That can be annoying to advanced users, but it is much more practical and for the majority of users probably the better default anyway.

Yes, we _could_ try to conserve byte-order/storage details more often.  And at least for `concatenate` that looks reasonable.  But unless someone can come up with a point where you have to deal with storage details, but can't pass `dtype=dtype_with_expected_storage`, I now think even there it is better to drop it.

(Even for `copy`, I am inclined that it might actually be a better default behaviour to "canonicalize".  Think of copying a structured dtype/void with a lot of unnecessary padding, removing the padding silently is good.  I am not actually suggesting we change that;  the argument is that even for `copy` preserving dtype storage details/byte-order is maybe not actually obvious?)

The issue itself is not about ufuncs silently dropping byte-order/storage details, but it has been mentioned.  This is would actually be tedious to get "right".  The ufunc machinery would have to:

1. Decide which output dtype to use for allocating the result array.
2. Decide which dtype to use for the inner-loop (i.e. cast the array to). (Which may now be different from the one in step 1.)

So for ufuncs trying to "fix" this, would actually require adding the _additional_ step 2, where currently we can use the same step for both.  If we only want to change it for `concatenate` it still means that we can't reuse `concatenate` logic for ufuncs, which also means duplicating almost identical paths.

---

Now, we could have went with a different world.  A world where `>float64` and `<float64` are considered distinct dtypes that interact seamlessly (i.e. storage details are important) – which also have some type of "equivalency" relation.  But that is not really where we are starting from:  `<float64` and `>float64` share a single scalar type, a single type-number and "character", and a single ufunc implementation. They are also equivalent from a static typing point of view (since we type this equivalently to scalar types right now).  In short, `<float64` and `<float64` are really both just one `Float64` type.

Basically, I think we have made the choice that byte-order (storage details) are not considered "user facing" when it comes to most users, but rather a detail that NumPy can modify as needed.  And I think its probably for the better.

**Comment by seberg** [comment]
@nschloe that comparison seems dubious to me:  You can't do `+a` as an operation on a big-endian array if you have a little-endian machine.  So yeah, you could special case `+a` to just copy the data, but that only works for `+` pretty much.  Also if  you would have more than one simple operation converting to native byte-order up-front would probably be the fasted to begin with (i.e. do an explicit `arr.astype()`) – unless your memory is larger than the CPU cache sizes anyway.

> There is no difference for very large arrays. (No idea why.)

Because the slow part for simple operation is dragging the whole array through the RAM into the CPU and out again.  The byte-swapping is just too fast compared to that.

**Comment by nschloe** [comment]
> You can't do +a as an operation on a big-endian array if you have a little-endian machine.

Aha, I thought the machine endianness was just the default for array, and that any machine could dp computation with big/little endian data. I see now that I was wrong, thanks for the explanation.

**Comment by seberg** [comment]
Stumbled on this again, and I am going to close it.  It seems to me the only _potential_ thing to be done here would be checking that the documentation is clear about dropping of byte-order during any form of promotion.
And in general, I think the number of users who need to know this is so small and the behaviour so clear (we always drop it), that this isn't even a super-high priority.

If anyone disagrees, please just comment.
